extern/boostd-data/couchbase/db.go

```diff
@@ -375,6 +420,23 @@ func (db *DB) GetOffsetSize(ctx context.Context, pieceCid cid.Cid, m multihash.M
 	return &model.OffsetSize{Offset: offset, Size: size}, nil
 }
 
+// Attempt to perform an update operation. If the operation fails due to a
+// cas mismatch, retry several times before giving up.
+func (db *DB) withCasRetry(f func() error) error {
+	var err error
+	for i := 0; i < maxCasRetries; i++ {
+		err = f()
+		if err == nil {
+			return nil
+		}
+		if !errors.Is(err, gocb.ErrCasMismatch) {
```

Comment:
I suggest as a start we add a log line here indicating that we got a CAS error.

Reply:
Agreed 👍 

I'm not expecting conflicts very often as the modification code paths are unlikely to be invoked simultaneously by multiple threads but I think it's a good idea to log CAS errors.

I'll add in a follow-up PR as this one is quite big already.